### ISSUE TITLE
[2022 ICLR Oral] BEIT: BERT Pre-Training of Image Transformers

### ISSUE THREAD

**Issue by Jasonlee1995** [Jasonlee1995/AI_Papers#65]
OpenAI의 DALL-E를 이용한 masked image modeling task를 통해 self-supervised vision representation을 얻음

## 1. Intuition

자연어에서는 많이 쓰이는 masked language modeling task를 이미지에 적용해보자는 취지에서 나옴

Denoising auto-encoding idea를 이용하여 vision transformer를 pretrain

자연어에서의 vocabulary를 image patch로 바꿈

Masked patches의 raw pixels를 예측하는 regression task로 설정할 수 있음

하지만 pixel-level recovery task는 모델이 short-range dependency와 high-frequency details를 학습하게 만듬
(DALL-E 참고)

이를 극복하기 위해 raw pixel을 예측하는 것이 아닌 visual token을 맞추는 task로 설정하여 pre-train

이를 통해 모델이 high-level abstractions를 학습하게 만들 수 있음

## 2. Bidirectional Encoder representation from Image Transformers (BEIT)
<img width="801" alt="image" src="https://user-images.githubusercontent.com/49643709/166260134-ab32fe33-0990-4a3f-adab-a6a6ce48acad.png">

### 2.1. Image Representations

Image는 image patch, visual tokens로 2가지 representation을 가짐

Image patch는 224 x 224 image를 14 x 14 grid의 image patches로 split함
(1개의 image patch는 16 x 16임)

Image를 14 x 14 grid의 visual tokens로 tokenize하며, 이는 pre-trained DALL-E를 통해 얻음
(vocabulary size = 8192)

즉 image patches 수와 visual tokens 수가 동일함

### 2.2. Pre-Training BEIT : Masked Image Modeling
<img width="761" alt="image" src="https://user-images.githubusercontent.com/49643709/166263285-5fea9493-6a0e-419c-b06d-2fa836d5166f.png">

Model의 output에 softmax classifier를 붙여 visual token을 predict

Corrupted image가 주어졌을 때 correct visual token의 log-likelihood를 maximize하는 방식으로 학습 진행

<img width="417" alt="image" src="https://user-images.githubusercontent.com/49643709/166263776-aea4a772-333a-4652-9fcb-0438cd6c9bbc.png">

기존의 자연어에서 사용되는 것과 비슷하게 blockwise masking을 통해 image를 masking

위 알고리즘을 통해 다양한 aspect ratio와 크기를 갖도록 mask를 생성함
(Algorithm의 h, w가 이미지의 세로, 가로가 아닌 image patch의 세로, 가로 개수임)

### 2.3. From the Perspective of Variational Autoencoder
<img width="755" alt="image" src="https://user-images.githubusercontent.com/49643709/166266169-bf3a7b22-7a08-4586-8f07-5088b3c27f65.png">

BEIT pre-training을 variational autoencoder training으로 볼 수 있음

Model을 2-stage procedure로 학습

1. Obtain the image tokenizer as a discrete variational autoencoder (tokenizer, tokenizer decoder)
2. Learn the prior while keeping tokenizer and tokenizer decoder fixed

여기서 tokenizer decoder이 one-point distribution으로 simplify하면 다음과 같이 식을 정리할 수 있음
(classification같이 가장 큰 값인 one point로 밀어버리면 kl-divergence가 nll로 바꿀 수 있음)

<img width="771" alt="image" src="https://user-images.githubusercontent.com/49643709/166267692-54e77b05-d102-4305-8360-7ec566840903.png">

### 3.1. Setup

- ViT-Base model 사용 : 12-layer Transformer with 768 hidden size, 12 attention heads, intermediate 3072 fc size 
- 16 x 16 input patch size
- Visual token vocabulary size : 8192
- Randomly mask at most 75 patches (roughly 40% of total image patches)

### 3.2. Fine-Tuning BEIT on Downstream Vision Tasks
<img width="741" alt="image" src="https://user-images.githubusercontent.com/49643709/166268659-7fb9f065-2666-4199-a333-08677f446cc3.png">

- Image classification : weight 곱한 후 average pooling + softmax를 통해 classification
- Semantic segmentation : use pre-trained BEIT as backbone encoder and incorporate several deconvolution layers as decoder
- Intermediate finetuning : 자연어에서 널리 쓰이는 방식과 같이 ImageNet-1K와 같은 data-rich intermediate dataset으로 학습한 후 target downstream tasks로 finetune

### 3.2. Image Classification
<img width="785" alt="image" src="https://user-images.githubusercontent.com/49643709/166269343-c23c013a-c003-4a79-9a05-5c1ea5a2ceba.png">
<img width="766" alt="image" src="https://user-images.githubusercontent.com/49643709/166270403-2b8caaa5-46b6-42c8-9856-20827c094317.png">
<img width="758" alt="image" src="https://user-images.githubusercontent.com/49643709/166280923-2b2f19dd-8ed7-403c-9daf-d055776233ea.png">

- Use larger learning rate with layer-wise decay
- Keep patch size the same for both 224 x 224 and 384 x 384 images
- After fine-tuning with resolution 224 x 224, additionally fine-tune the model on 384 x 384 images
- Benefits of scaling up BEIT from base to large are greater than supervised pre-training
- BEIT tends to help more for extremely larger models
- Fine-tuning BEIT not only achieves better performance, but also converging much faster

### 3.3. Semantic Segmentation
<img width="773" alt="image" src="https://user-images.githubusercontent.com/49643709/166282850-10bb64ad-29d8-464b-8401-ddced5e587c1.png">

### 3.4. Ablation Studies
<img width="756" alt="image" src="https://user-images.githubusercontent.com/49643709/166286401-694b25a8-56c8-47eb-855c-ee874e355b30.png">

- Blockwise masking is beneficial on both tasks, especially on semantic segmentation
- Predicting visual tokens > predicting the raw pixels of masked patches
- Blockwise masking is even more helpful for pixel-level auto-encoding, which relieves the suffering of short-distance dependency
- Pre-training the model longer can further improve performance on downstream tasks

### 3.5. Analysis of Self-Attention Map
<img width="765" alt="image" src="https://user-images.githubusercontent.com/49643709/166289874-4e31d7ca-1ad1-474b-8393-72d3abb28b6d.png">

### 3.6. Multi-Task Pre-Training with DINO
<img width="762" alt="image" src="https://user-images.githubusercontent.com/49643709/166290034-82981442-3a5c-4b65-b84f-38993fd8cb4d.png">

BEIT와 DINO를 multi-task manner로 같이 pre-train하면 semantic segmentation 성능이 올라감
